### DISCUSSION TITLE
Question about SPI drivers

### DISCUSSION THREAD

**Discussion by imechura** [grblHAL/core#296] [General]
Hello,

Are there any drivers for GRBLHal that support using SPI to control stepper movement? I am trying to use GRBLHal with a trinamic TMCM-301 motion conroller which does not accept step/dir signals. Instead it requires sending SPI datagrams in steps that represent a sinusoidal waveform. I have seen trinamic code that is used to configure the drivers via SPI bt I have not been able to find any code that controls the motor position that way.

Thanks,

**Comment by terjeio** [comment]
It is _possible_ to drive steppers that way as the current position is available to the driver code. But I do not think it is _feasible_ since it might be hard to achive syncronized motion among axes.

Anyway, the core stepper code typically splits motion up in approx. 5ms segments during which direction and step rate does not change. You may try to send data to the controllers for each new [segment](https://github.com/grblHAL/core/blob/339cfd5e817c5333f681b5c4b97e9952dbb54495/stepper.h#L57-L72) as a starting point?

**Reply by imechura** [reply]
terjeio,

I have not yet developed the domain knowlege that would be required to intelligently respond to your message. Ive been programming for a while but GRBLHal and machine control are relativley new concepts that I am still learning about.

I do not yet uderstand the process that grblhal uses to transform gcode directives into pulses (assuming there is a such a thing). Could you refer me to the a starting point in the code so I can read it an attempt to gain a better understanding on the subject?

I have read through some of the code in these repositories, Ive understool a little bit, but I 'm not sure I have come across that part.
Thanks,

**Reply by terjeio** [reply]
[This article](https://awesome.tech/grbl-demystified/) is helpful?

[mc_line()](https://github.com/grblHAL/core/blob/339cfd5e817c5333f681b5c4b97e9952dbb54495/motion_control.c#L71-L78) is where motion is submitted for execution.
